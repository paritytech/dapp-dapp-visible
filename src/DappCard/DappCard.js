// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React, { Component } from 'react';
import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import Card from 'semantic-ui-react/dist/commonjs/views/Card';
import Button from 'semantic-ui-react/dist/commonjs/elements/Button';
import Label from 'semantic-ui-react/dist/commonjs/elements/Label';
import DappIcon from '@parity/ui/lib/DappIcon';
import { FormattedMessage } from 'react-intl';

import DappVouchFor from './DappVouchFor';
import styles from './DappCard.css';

const DAPP_DAPP_VISIBLE_ID =
  '0xa48bd8fd56c90c899135281967a6cf90865c221b46f27f9fbe2a236d74a64ea2'; // This dapp's id

class DappCard extends Component {
  static propTypes = {
    dapp: PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      image: PropTypes.string
    }).isRequired,
    onAdd: PropTypes.func.isRequired,
    onOpen: PropTypes.func.isRequired,
    onRemove: PropTypes.func.isRequired,
    visible: PropTypes.bool
  };

  handleAdd = () => this.props.onAdd(this.props.dapp.id);

  handleRemove = () => this.props.onRemove(this.props.dapp.id);

  handleOpen = () => this.props.onOpen(this.props.dapp.id);

  render() {
    const { dapp, visible } = this.props;

    return (
      <Card>
        <Card.Content>
          <DappIcon
            app={dapp}
            className={`ui centered image ${styles.picture}`}
          />
          <Card.Header>{dapp.name}</Card.Header>
          <Card.Meta>{dapp.description}</Card.Meta>
          <Card.Description>
            {dapp.type === 'local' && (
              <Label>
                <FormattedMessage
                  id="dapp.visible.localLabel"
                  defaultMessage="Local Dapp"
                />
              </Label>
            )}
            {dapp.type === 'builtin' && (
              <Label>
                <FormattedMessage
                  id="dapp.visible.builtinLabel"
                  defaultMessage="By Parity"
                />
              </Label>
            )}
            <DappVouchFor dapp={dapp} />
          </Card.Description>
        </Card.Content>
        <Card.Content extra>
          <Button
            basic
            secondary
            content={
              <FormattedMessage
                id="dapp.visible.openDapp"
                defaultMessage="Open"
              />
            }
            icon="external"
            onClick={this.handleOpen}
          />
          {visible ? (
            <Button
              content={
                <FormattedMessage
                  id="dapp.visible.addedDapp"
                  defaultMessage="Added to home"
                />
              }
              disabled={dapp.id === DAPP_DAPP_VISIBLE_ID} // Do not allow removing this dapp from homepage
              icon="toggle on"
              onClick={this.handleRemove}
              primary
            />
          ) : (
            <Button
              basic
              content={
                <FormattedMessage
                  id="dapp.visible.addDapp"
                  defaultMessage="Add to home"
                />
              }
              icon="toggle off"
              onClick={this.handleAdd}
              primary
            />
          )}
        </Card.Content>
      </Card>
    );
  }
}

export default observer(DappCard);
