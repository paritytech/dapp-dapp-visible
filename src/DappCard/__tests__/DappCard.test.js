// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { shallowToJson } from 'enzyme-to-json';

import { shallowWithIntl, mountWithIntl } from '../../setupTests';
import DappCard from '../DappCard';

const props = {
  dapp: {
    id: '123',
    name: 'Test'
  },
  methodGroups: { shell: { methods: ['foo', 'bar'] } },
  permissions: { 'foo:123': true },
  onEdit: () => {},
  onToggle: () => {}
};

test('should render correctly in viewing mode', () => {
  const component = shallowWithIntl(<DappCard {...props} />);

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should render correctly in viewing mode with no allowed methods', () => {
  const component = shallowWithIntl(<DappCard {...props} permissions={{}} />);

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should render correctly in editing mode', () => {
  const component = shallowWithIntl(<DappCard {...props} editingMode />);

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should handle onEdit button click', () => {
  const onEdit = jest.fn();

  const component = shallowWithIntl(<DappCard {...props} onEdit={onEdit} />);

  component
    .find(Button)
    .last()
    .simulate('click');
  expect(onEdit).toHaveBeenCalled();
});

test('should handle onToggle click', () => {
  const onToggle = jest.fn();

  const component = mountWithIntl(
    <DappCard {...props} editingMode onToggle={onToggle} />
  );

  component
    .find(List.Item)
    .last() // The 'bar' one
    .simulate('click');

  expect(onToggle).toHaveBeenCalledWith('bar', '123');
});
