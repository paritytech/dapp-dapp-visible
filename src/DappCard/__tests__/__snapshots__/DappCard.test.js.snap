// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly in editing mode 1`] = `
<Card>
  <CardContent>
    <Button
      as="button"
      basic={true}
      floated="right"
      icon="remove"
      onClick={[Function]}
      size="mini"
    />
    <Image
      as="img"
      centered={true}
      ui={true}
    />
    <CardHeader>
      Test
    </CardHeader>
    <CardMeta />
    <CardDescription>
      <Accordion
        panels={
          Array [
            Object {
              "content": Object {
                "active": true,
                "content": <List
                  className={undefined}
          >
                  <ListItem
                          className={undefined}
                          onClick={[Function]}
                  >
                          <ListIcon
                                  name="checkmark box"
                          />
                          foo
                  </ListItem>
                  <ListItem
                          className={undefined}
                          onClick={[Function]}
                  >
                          <ListIcon
                                  name="square outline"
                          />
                          bar
                  </ListItem>
          </List>,
                "key": "123-shell-editing-content",
              },
              "title": Object {
                "active": true,
                "content": <FormattedMessage
                  defaultMessage="Can get access to the Shell"
                  description="Explanation of the shell methodGroup permission"
                  id="dapp.methods.shell"
                  values={Object {}}
          />,
                "key": "123-shell-editing-title",
              },
            },
          ]
        }
      />
    </CardDescription>
  </CardContent>
</Card>
`;

exports[`should render correctly in viewing mode 1`] = `
<Card>
  <CardContent>
    <Button
      as="button"
      basic={true}
      floated="right"
      icon="edit"
      onClick={[Function]}
      size="mini"
    />
    <Image
      as="img"
      centered={true}
      ui={true}
    />
    <CardHeader>
      Test
    </CardHeader>
    <CardMeta />
    <CardDescription>
      <Accordion
        panels={
          Array [
            Object {
              "content": Object {
                "content": <List
                  bulleted={true}
                  className={undefined}
          >
                  <ListItem>
                          foo
                  </ListItem>
          </List>,
                "key": "123-shell-content",
              },
              "title": Object {
                "content": <FormattedMessage
                  defaultMessage="Can get access to the Shell"
                  description="Explanation of the shell methodGroup permission"
                  id="dapp.methods.shell"
                  values={Object {}}
          />,
                "key": "123-shell-title",
              },
            },
          ]
        }
      />
    </CardDescription>
  </CardContent>
</Card>
`;

exports[`should render correctly in viewing mode with no allowed methods 1`] = `
<Card>
  <CardContent>
    <Button
      as="button"
      basic={true}
      floated="right"
      icon="edit"
      onClick={[Function]}
      size="mini"
    />
    <Image
      as="img"
      centered={true}
      ui={true}
    />
    <CardHeader>
      Test
    </CardHeader>
    <CardMeta />
    <CardDescription>
      <div>
        <FormattedMessage
          defaultMessage="No allowed methods"
          id="dapps.methods.noAllowedMethods"
          values={Object {}}
        />
      </div>
    </CardDescription>
  </CardContent>
</Card>
`;
